queries/highlights.scm

```diff
@@ -162,15 +168,244 @@
+] @repeat
+(for "in" @repeat)
+; keywords related to loops (e.g. `for` / `while`)
+
+[
+  "block"
+] @label
+; GOTO and other labels (e.g. `label:` in C)
+
+[
+  "import"
+  "include"
+  "export"
+] @include
+(import_from_statement "from" @include)
+(except_clause "except" @include)
+; keywords for including modules (e.g. `import` / `from` in Python)
+
+[
+  "try"
+  "except"
+  "finally"
+  "raise"
+] @exception
+; keywords related to exceptions (e.g. `throw` / `catch`)
+
+((generic_parameter_list
+  (parameter_declaration
+    (symbol_declaration_list
+      (symbol_declaration name: (_) @type))))
+ (#set! "priority" 101)) ; > @parameter
+
+(call
+  (bracket_expression 
+    right: (argument_list [
+      (identifier) @type 
+      (accent_quoted (identifier) @type)
+    ]))
+)
+; match generic types when calling 
+; FIX: this also falsy matches when accessing and call elements from an array of routines
+
+(type_symbol_declaration name: (_) @type)
+
+(type_expression) @type
+
+(infix_expression operator: "is" right: (_) @type)
+
+; TODO: is there another way?
+(except_branch values: (expression_list
+  [
+    (identifier) @type
+    (infix_expression
+      left: (_) @type
+      operator: "as"
+      right: (identifier) @variable)
+    (_
+      (infix_expression
+        left: (_) @type
+        operator: "as"
+        right: (identifier) @variable))
+    (_ (_
+        (infix_expression
+          left: (_) @type
+          operator: "as"
+          right: (identifier) @variable)))
+    (_ (_ (_
+        (infix_expression
+          left: (_) @type
+          operator: "as"
+          right: (identifier) @variable))))
+  ]))
+
+; type or class definitions and annotations
+
+((tuple_type "tuple" @type.builtin (field_declaration_list))
+ (#set! "priority" 101)) ; > @type
+; for inline tuple types
+
+; TODO: is there another way?
+(
+ [
+   (type_expression (identifier) @type.builtin) 
+   (type_expression (_ (identifier) @type.builtin)) 
+   (type_expression (_ (_ (identifier) @type.builtin))) 
+   (type_expression (_ (_ (_ (identifier) @type.builtin)))) 
+   (type_expression (_ (_ (_ (_ (identifier) @type.builtin))))) 
+   (type_expression (_ (_ (_ (_ (_ (identifier) @type.builtin)))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin))))))))))))) 
+   (type_expression (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (_ (identifier) @type.builtin)))))))))))))) 
+ ]
+ (#set! "priority" 101) ; > @type
+ (#any-of? @type.builtin
```

Comment:
surely the lsp can do a way better job with all the semantic dependent highlighting. Same with the constructors, method calls, constants, enum and some other corner cases. Does the nim lsp already support semantic tokens?

Reply:
It doesn't but I'd argue that it's a lot more suited for this purpose. `nimsuggest` can already do it (my nvim plugin uses that), so there's no reason why the LSP implementations can't, they are already using nimsuggest under the hood.

I'd prefer not to fill the queries with rules like this when an LSP can perform a much better job.

Reply:
I get your meaning. I think for this specific case it makes sense, since the builtin types don't change and it will improve the experience for the user IMO. Also I tested the performance impact and it was minor.

`$ for i in {1..10}; do tree-sitter query -t -q highlights.scm $NimRepo/tests/manyloc/keineschweine/lib/glext.nim; done`

without @type.builtin:
```
28.597137ms
28.587892ms
28.805097ms
28.720249ms
28.570727ms
28.524449ms
28.543346ms
28.555084ms
28.611419ms
28.982347ms
```
with @type.builtin:
```
28.807002ms
29.598144ms
28.941699ms
28.811389ms
28.78843ms
29.056218ms
29.020567ms
28.839141ms
29.020161ms
28.869629ms
```

Not the most thorough test, but you get the gist.

Reply:
There is also the option to make some weak heuristic like built-in types are usually lowercase, while user types are supposed to be PascalCase, but really that's just a heuristic and the user can do what they want.
Could also make similar heuristics for constants, enums, module names etc., but I don't think it's great.

Reply:
I worry more about the unwieldy-ness of the grammar.

I'd say we need test cases for these things.

Reply:
> I worry more about the unwieldy-ness of the grammar.

you mean the grammar or the highlight queries?

> I'd say we need test cases for these things.
Sure, but writing test cases for queries suck. writing that syntax is really time intesive.